origin_ui/origin_api.go

```diff
@@ -0,0 +1,189 @@
+	authHeader := ctx.Request.Header.Get("Authorization")
+
+	// Check if the Authorization header was provided
+	if authHeader == "" {
+		// Use AbortWithStatusJSON to stop invoking the next chain
+		ctx.AbortWithStatusJSON(401, gin.H{"error": "Authorization header is missing"})
+		return
+	}
+
+	// Check if the Authorization type is Bearer
+	if !strings.HasPrefix(authHeader, "Bearer ") {
+		ctx.AbortWithStatusJSON(401, gin.H{"error": "Authorization header is not Bearer type"})
+		return
+	}
+
+	// Extract the token from the Authorization header
+	token := strings.TrimPrefix(authHeader, "Bearer ")
+	valid, err := director.VerifyDirectorTestReportToken(token)
+
+	if err != nil {
+		log.Warningln(fmt.Sprintf("Error when verifying Bearer token: %s", err))
+		ctx.AbortWithStatusJSON(401, gin.H{"error": fmt.Sprintf("Error when verifying Bearer token: %s", err)})
+		return
+	}
+
+	if !valid {
+		log.Warningln("Can't validate Bearer token")
+		ctx.AbortWithStatusJSON(401, gin.H{"error": "Can't validate Bearer token"})
+		return
+	}
+	ctx.Next()
+}
+
+// Reset the timer safely
+func resetDirectorTimeoutTimer() {
+	timerMutex.Lock()
+	defer timerMutex.Unlock()
+
+	if directorTimeoutTimer == nil {
+		directorTimeoutTimer = time.NewTimer(directorTimeoutDuration)
+		go func() {
+			for {
+				select {
+				case <-directorTimeoutTimer.C:
+					// Timer fired because no message was received in time.
+					log.Warningln("No director test report received within the time limit")
+					if err := metrics.SetComponentHealthStatus("director", "critical", "No director test report received within the time limit"); err != nil {
+						log.Errorln("Failed to update director component health status:", err)
+					}
+					// Reset the timer for the next period.
+					timerMutex.Lock()
+					directorTimeoutTimer.Reset(directorTimeoutDuration)
+					timerMutex.Unlock()
+				case <-exitLoop:
+					log.Infoln("Gracefully terminating the director-health test timeout loop...")
+					return
+				}
+			}
+		}()
+	} else {
+		if !directorTimeoutTimer.Stop() {
+			<-directorTimeoutTimer.C
+		}
+		directorTimeoutTimer.Reset(directorTimeoutDuration)
+	}
+}
+
+// Director will periodiclly upload/download files to/from all connected
+// origins and test the health status of origins. It will send a request
+// reporting such status to this endpoint, and we will update origin internal
+// health status metric to reflect the director connection status.
+func directorTestResponse(ctx *gin.Context) {
+	dt := DirectorTest{}
+	if err := ctx.ShouldBind(&dt); err != nil {
+		log.Errorf("Invalid director test response")
+		ctx.JSON(400, gin.H{"error": "Invalid director test response"})
+		return
+	}
+	// We will let the timer go timeout if director didn't send a valid json request
+	resetDirectorTimeoutTimer()
+	if dt.Status == "ok" {
+		if err := metrics.SetComponentHealthStatus("director", "ok", fmt.Sprintf("Director timestamp: %s", dt.Timestamp)); err != nil {
```

Comment:
For future work -- I hate all the strings in this interface.  We should cast it to an enum.

IIRC, the reason why `SetComponentHealthStatus` returns an error is because it internally casts to an enum.  If we forced type safety on the caller, then we could eliminate the error return and simplify the code base.